src/gt4py/next/program_processors/runners/dace_fieldview/gtir_to_sdfg.py

```diff
@@ -324,38 +324,86 @@ def visit_FunCall(
         node: gtir.FunCall,
         sdfg: dace.SDFG,
         head_state: dace.SDFGState,
+        let_symbols: dict[str, gtir_builtin_translators.TemporaryData],
     ) -> list[gtir_builtin_translators.TemporaryData]:
         # use specialized dataflow builder classes for each builtin function
         if cpm.is_call_to(node, "cond"):
-            return gtir_builtin_translators.translate_cond(node, sdfg, head_state, self)
+            return gtir_builtin_translators.translate_cond(
+                node, sdfg, head_state, self, let_symbols
+            )
         elif cpm.is_call_to(node.fun, "as_fieldop"):
-            return gtir_builtin_translators.translate_as_field_op(node, sdfg, head_state, self)
+            return gtir_builtin_translators.translate_as_field_op(
+                node, sdfg, head_state, self, let_symbols
+            )
+        elif isinstance(node.fun, gtir.Lambda):
+            node_args = []
+            for arg in node.args:
+                node_args.extend(
+                    self.visit(
+                        arg,
+                        sdfg=sdfg,
+                        head_state=head_state,
+                        let_symbols=let_symbols,
+                    )
+                )
+            return self.visit(
+                node.fun,
+                sdfg=sdfg,
+                head_state=head_state,
+                let_symbols=let_symbols,
+                args=node_args,
+            )
         else:
             raise NotImplementedError(f"Unexpected 'FunCall' expression ({node}).")
 
-    def visit_Lambda(self, node: gtir.Lambda) -> Any:
+    def visit_Lambda(
+        self,
+        node: gtir.Lambda,
+        sdfg: dace.SDFG,
+        head_state: dace.SDFGState,
+        let_symbols: dict[str, gtir_builtin_translators.TemporaryData],
+        args: list[gtir_builtin_translators.TemporaryData],
+    ) -> list[gtir_builtin_translators.TemporaryData]:
         """
-        This visitor class should never encounter `itir.Lambda` expressions
-        because a lambda represents a stencil, which operates from iterator to values.
-        In fieldview, lambdas should only be arguments to field operators (`as_field_op`).
+        Translates a `Lambda` node to a tasklet subgraph in the current SDFG state.
+
+        All arguments to lambda functions are fields (i.e. `as_fieldop`, field or scalar `gtir.SymRef`,
+        nested let-lambdas thereof). The dictionary called `let_symbols` maps the lambda parameters
+        to symbols, e.g. temporary fields or program arguments. The lambda parameters can override
```

Comment:
I would write something like. "If the Lambda has an argument whose name is a (previously defined) let symbols, i.e. is also mentioned in `let_symbols` the argument will, during the traversal of the lambda expression, shadow that let symbol."

I also want to conform one thing, this location is the only place where the `let_symbols` get modified?
Furthermore, I am unsure about the behaviour, as I understand it, the current implementation makes all arguments the lambda have available to all of its children, regardless how remote they are and what is between them (with the exception of another Lambda that has the same argument name)?
For this to happen it is not required that a let expression was there in the first place.
I am not sure if this is the intended behaviour or not I just want to confirm it.

My main issue is that it makes the term "let-lambda" meaningless, as every lambda has essentially the same possibilities. On the other hand I see nothing in the IR that allows to create a special let lambda.

Furthermore, another question I had was, if you add all arguments of the lambda to the `let_symbols` are they considered at all?

Reply:
I will update the docstring as you suggested, here I try to answer your questions.

- Yes, this is the only location where we can override (shadow) previously defined symbols.
- Yes, the current implementation makes all arguments the lambda have available to all of its children. Not only that, a lambda can also access all "global" symbols, that is the program arguments. This behavior can be described, in other words, as the lambda inherits the symbols in the parent scope. I observed this behavior was required in ITIR, I assume it is still valid here @tehrengruber? I mean, a lambda can access symbols defined in the parent scope, not only its own parameters.
- The term "let-lambda" is used to distinguish the regular usage of lambda functions in GTIR (kind of stencils passed as argument to `as_fieldop`) from this particular lambda that creates an intermediate result (a temporary field, indeed) in the current scope and makes it available to the expressions in the child nodes.
- I am not sure I understand your last question. Yes, they are considered because `translate_symbol_ref` first checks if the symbol is present in `let_symbols`

Reply:
Thanks for your answer.
I was kind of expecting that answer, but thanks for conformation.

Concerning my last point, what I meant by it, since all arguments of a lambda are added to the `let_symbols` will the argument properties of the lambda ever by considered, since you do not have to look at it, since it was evaluated before.
This is more of a question.

Reply:
I have now pushed a solution for the problem we discussed on Zoom. I opted for the separate lambda-symbols state, because the direction of SDFG growth (adding states before the head state) is designed for the regular traversal order. The case I showed to you, with let-lambdas as parent of conditional expressions (for which we thought that the conditional node should modify the head state), is specific to let-lambdas and I consider it an exception.
To answer your question: first (in `visit_FunCall`) we evaluate the arguments passed to lambda, and lower it to a dataflow in a separate state that precedes lambda execution; then (in `visit_Lambda`), we map these arguments to the lambda parameters and visit the lambda expression. The lambda expression will contain `SymRef` to the lambda parameters, and at that point we add access nodes to the arguments.
The approach I opted for - the separate lambda-symbols state - relies on the simplify pass to remove redundant states.